src/pooler.c

```diff
@@ -57,6 +57,47 @@ static struct timeval err_timeout = {5, 0};
 
 static void tune_accept(int sock, bool on);
 
+static void close_socket(struct ListenSocket *ls)
+{
+	if (event_del(&ls->ev) < 0) {
+		log_warning("close_socket, event_del: %s", strerror(errno));
+	}
+	if (ls->fd > 0) {
+		safe_close(ls->fd);
+		ls->fd = 0;
+	}
+	if (pga_is_unix(&ls->addr) && cf_unix_socket_dir[0] != '@') {
+		char buf[sizeof(struct sockaddr_un) + 20];
+		snprintf(buf, sizeof(buf), "%s/.s.PGSQL.%d", cf_unix_socket_dir, cf_listen_port);
+		unlink(buf);
+	}
+	statlist_remove(&sock_list, &ls->node);
```

Comment:
This is a bit weird. Refactoring this function out from `cleanup_sockets()` makes sense, but then this should only work on a single socket in isolation, not manipulate a higher-layer list.